sdk/servicebus/service-bus/src/core/shared.ts

```diff
@@ -103,3 +106,132 @@ export function createReceiverOptions(
 
   return rcvrOptions;
 }
+
+export const UnsettledMessagesLimitExceededError =
+  "Failed to fetch new messages as the limit for unsettled messages is reached. Please settle received messages using settlement methods(such as `completeMessage()`) on the receiver to receive the next message.";
+
+/**
+ * Returns the number of empty/filled slots in the Circular buffer of incoming deliveries
+ * based on the capacity and size of the buffer.
+ *
+ * @internal
+ */
+export function incomingBufferProperties(
+  receiver: Pick<Receiver, "session"> | undefined
+): {
+  numberOfEmptySlots: number; // 2048(total) - filled
+  numberOfFilledSlots: number; // number of unsettled messages
+} {
+  const incomingDeliveries = receiver?.session?.incoming?.deliveries;
+  let numberOfEmptySlots = 0;
+  if (incomingDeliveries && incomingDeliveries.capacity - 1 > incomingDeliveries.size) {
+    // Exmpty slots should have been `incomingDeliveries.capacity - 1 - incomingDeliveries.size`. Why -1?
+    // - If the number of slots is set to (capacity - size),
+    //   the number of unsettled messages that can be held in the buffer would equal to the "capacity".
+    //   At that limiting point, service doesn't respond to the drain request for unpartitioned queues.
+    //   Service team is tracking the issue.
+    // -1 allows us to not fill up the buffer entirely, it fills up to 2047 if the capacity is 2048
+    numberOfEmptySlots = incomingDeliveries.capacity - 1 - incomingDeliveries.size;
+  }
+  return {
+    numberOfEmptySlots,
+    numberOfFilledSlots: incomingDeliveries ? incomingDeliveries.size : 0
+  };
+}
+
+/**
+ * Provides helper methods to manage the credits on the link for the
+ * streaming messages scenarios.
+ * (Used by both sessions(MessageSession) and non-sessions(StreamingReceiver))
+ *
+ * @internal
+ */
+export class StreamingReceiverCreditManager {
+  constructor(
+    private _getCurrentReceiver: () => { receiver: Receiver | undefined; logPrefix: string },
```

Comment:
Not actually.
I copied this part of code from the StreamingReceiverHelper. 
Once I checked(after your comment), I realized that StreamingReceiverHelper has guards to make sure the receiver not defined.

Instead of redoing, I've merged my StreamingReceiverCreditManager into the existing StreamingReceiverHelper.

Let me know if that looks fine.